[AIRFLOW-XXX] GSoD: Adding Task re-run documentation (#6295)
mistercrunch/Airflow: docs/dag-run.rst

..  Licensed to the Apache Software Foundation (ASF) under one
    or more contributor license agreements.  See the NOTICE file
    distributed with this work for additional information
    regarding copyright ownership.  The ASF licenses this file
    to you under the Apache License, Version 2.0 (the
    "License"); you may not use this file except in compliance
    with the License.  You may obtain a copy of the License at

..    http://www.apache.org/licenses/LICENSE-2.0

..  Unless required by applicable law or agreed to in writing,
    software distributed under the License is distributed on an
    "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
    KIND, either express or implied.  See the License for the
    specific language governing permissions and limitations
    under the License.

DAG Runs
=========

> A DAG Run is an object representing an instantiation of the DAG in time

Reviewer: ```suggestion
A DAG Run is an object representing an instantiation of the DAG in time.
```

| ``@monthly`` | Run once a month at midnight of the first day of the month     | ``0 0 1 * *`` |
+--------------+----------------------------------------------------------------+---------------+
| ``@yearly``  | Run once a year at midnight of January 1                       | ``0 0 1 1 *`` |
+--------------+----------------------------------------------------------------+---------------+

Your DAG will be instantiated for each schedule along with a corresponding 
``DAG Run`` entry in backend.

**Note**: If you run a DAG on a schedule_interval of one day, the run stamped 2020-01-01 
will be triggered soon after 2020-01-01T23:59. In other words, the job instance is 
started once the period it covers has ended.  The execution_date passed in the dag 
will also be 2020-01-01.

The first ``DAG Run`` is created based on the minimum ``start_date`` for the tasks in your DAG. 
Subsequent ``DAG Runs`` are created by the scheduler process, based on your DAG’s ``schedule_interval``, 
sequentially. If your start_date is 2020-01-01 and schedule_interval is @daily the first run 
will be created on 2020-01-02 i.e. after your start date has passed.

Re-run DAG
''''''''''

> 

Reviewer: Can you add a small introduction here?

.. code:: python

    """
    Code that goes along with the Airflow tutorial located at:
    https://github.com/apache/airflow/blob/master/airflow/example_dags/tutorial.py
    """
    from airflow import DAG
    from airflow.operators.bash_operator import BashOperator
    from datetime import datetime, timedelta

    default_args = {
        'owner': 'Airflow',
        'depends_on_past': False,
        'start_date': datetime(2015, 12, 1),
        'email': ['airflow@example.com'],
        'email_on_failure': False,
        'email_on_retry': False,
        'retries': 1,
        'retry_delay': timedelta(minutes=5)
    }

    dag = DAG(
        'tutorial',
        default_args=default_args,
        description='A simple tutorial DAG',
        schedule_interval='@daily',
        catchup=False)

In the example above, if the DAG is picked up by the scheduler daemon on 2016-01-02 at 6 AM, 
(or from the command line), a single DAG Run will be created, with an `execution_date` of 2016-01-01, 
and the next one will be created just after midnight on the morning of 2016-01-03 with an execution date of 2016-01-02.

> If the `dag.catchup` value had been True instead, the scheduler would have created a DAG Run 

Reviewer: ```suggestion
If the ``dag.catchup`` value had been True instead, the scheduler would have created a DAG Run 
```

for each completed interval between 2015-12-01 and 2016-01-02 (but not yet one for 2016-01-02, 

> as that interval hasn’t completed) and the scheduler will execute them sequentially. This behavior 

Reviewer: ```suggestion
as that interval hasn’t completed) and the scheduler will execute them sequentially. 

This behavior 
```
This is a very important summary sentence, which should be included in a separate paragraph.

is great for atomic datasets that can easily be split into periods. Turning catchup off is great 
if your DAG Runs perform backfill internally.

> Backfill

Reviewer: Can you explain that word?

---------

You may want to backfill the data even in the cases when catchup is disabled. This can be done through CLI. 
Run the below command

.. code:: bash

    airflow backfill -s START_DATE -e END_DATE dag_id

The above command will re-run all the instances of the dag_id for all the intervals within the start date and end date.

Re-run Tasks
------------
It can happen, that some of the tasks can fail during the scheduled run. Once you have fixed 
the errors after going through the logs, you can re-run the tasks by clearing it for the 
scheduled date. Clearing a task instance doesn't delete the task instance record. 

> Instead it updates max_tries and set the current task instance state to be None.

Reviewer: ```suggestion
Instead it updates ``max_tries`` to ``0`` and set the current task instance state to be ``None``, this forces the task to re-run.
```

> Select the failed task and click on Clear. This will clear the status of the task from 

Reviewer: ```suggestion
Select the failed task and click on **Clear**. This will clear the status of the task from 
```

> failed to None and the executor will re-run it.

Reviewer: ```suggestion
failed to ``None`` and the executor will re-run it.
```

There are multiple options you can select to re-run - 

* Past - All the instances of the task in the  runs before the current DAG's execution date
* Future -  All the instances of the task in the  runs after the current DAG's execution date
* Upstream - The upstream tasks in the current DAG
* Downstream - The downstream tasks in the current DAG
* Recursive - All the tasks in the child DAGs and parent DAGs
* Failed - Only the failed tasks in the current DAG

> You can also clear the task through CLI using the command -

Reviewer: ```suggestion
You can also clear the task through CLI using the command:
```

.. code:: bash

    airflow tasks clear dag_id -t task_regex -s START_DATE -d END_DATE

This will clear all instances of the tasks matching the regex for the dag_id which have run during 

> the mentioned interval. For more options, you can run the command :

Reviewer: ```suggestion
the mentioned interval. For more options, you can run the command:
```

.. code:: bash

    airflow tasks clear -h

**Note**: When clearing a set of tasks’ state in hope of getting them to re-run, it is important 

> to keep in mind the `DAG Run`’s state too as it defines whether the scheduler should look

Reviewer: ```suggestion
to keep in mind the DAG Run’s state too as it defines whether the scheduler should look
```

into triggering tasks for that run.

External Triggers
'''''''''''''''''

Note that ``DAG Runs`` can also be created manually through the CLI. Just run the command -

.. code:: bash

    airflow dags trigger -e execution_date run_id

The ``DAG Runs`` created externally to the scheduler get associated to the trigger’s timestamp, and will be displayed 
in the UI alongside scheduled ``DAG runs``. The execution date passed inside the DAG can be specified using -e argument 
otherwise the current date in UTC timezone is used as default.

> In addition, you can also manually trigger a `DAG Run` using the web UI (tab “DAGs” -> column “Links” -> button “Trigger Dag”)

Reviewer: ```suggestion
In addition, you can also manually trigger a `DAG Run` using the web UI (tab **DAGs** -> column **Links** -> button **Trigger Dag**)
```

+--------------+----------------------------------------------------------------+---------------+
| ``@hourly``  | Run once an hour at the beginning of the hour                  | ``0 * * * *`` |
+--------------+----------------------------------------------------------------+---------------+
| ``@daily``   | Run once a day at midnight                                     | ``0 0 * * *`` |
+--------------+----------------------------------------------------------------+---------------+
| ``@weekly``  | Run once a week at midnight on Sunday morning                  | ``0 0 * * 0`` |
+--------------+----------------------------------------------------------------+---------------+
| ``@monthly`` | Run once a month at midnight of the first day of the month     | ``0 0 1 * *`` |
+--------------+----------------------------------------------------------------+---------------+
| ``@yearly``  | Run once a year at midnight of January 1                       | ``0 0 1 1 *`` |
+--------------+----------------------------------------------------------------+---------------+

Your DAG will be instantiated for each schedule along with a corresponding 
DAG Run entry in backend.

> **Note**: If you run a DAG on a schedule_interval of one day, the run stamped 2020-01-01 

Reviewer 2: I think this and the next paragraph sould be in a `.. note::` block to make them stand out more. What does that look like?

 .. Licensed to the Apache Software Foundation (ASF) under one
    or more contributor license agreements.  See the NOTICE file
    distributed with this work for additional information
    regarding copyright ownership.  The ASF licenses this file
    to you under the Apache License, Version 2.0 (the
    "License"); you may not use this file except in compliance
    with the License.  You may obtain a copy of the License at

 ..   http://www.apache.org/licenses/LICENSE-2.0

 .. Unless required by applicable law or agreed to in writing,
    software distributed under the License is distributed on an
    "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
    KIND, either express or implied.  See the License for the
    specific language governing permissions and limitations
    under the License.

DAG Runs
=========
A DAG Run is an object representing an instantiation of the DAG in time.

Each DAG may or may not have a schedule, which informs how DAG Runs are
created. ``schedule_interval`` is defined as a DAG argument, and receives
preferably a

> `cron expression <https://en.wikipedia.org/wiki/Cron#CRON_expression>`_ as

Reviewer 3: I know it might be biased to refer to certain projects, but it might be nice to reference [crontab.guru](https://crontab.guru/) as it helps people create crons and displays the period at which they would run.

Reviewer 4: Great idea, but probably best to link to crontab.guru in the wikipedia page itself.

EDIT: Never mind. This is non-trivial / probably not worth it as `*.guru` links are blacklisted in wikipedia 🤷‍♂ - unless someone's a Wikipedia guru, it seems like a hassle to unblock it.

Reviewer 3: 😢

Reviewer 2: I don't see a problem linking to crontab.guru as well in an aside.

 .. Licensed to the Apache Software Foundation (ASF) under one
    or more contributor license agreements.  See the NOTICE file
    distributed with this work for additional information
    regarding copyright ownership.  The ASF licenses this file
    to you under the Apache License, Version 2.0 (the
    "License"); you may not use this file except in compliance
    with the License.  You may obtain a copy of the License at

 ..   http://www.apache.org/licenses/LICENSE-2.0

 .. Unless required by applicable law or agreed to in writing,
    software distributed under the License is distributed on an
    "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
    KIND, either express or implied.  See the License for the
    specific language governing permissions and limitations
    under the License.

DAG Runs
=========
A DAG Run is an object representing an instantiation of the DAG in time.

Each DAG may or may not have a schedule, which informs how ``DAG Runs`` are

> created. ``schedule_interval`` is defined as a DAG arguments, and receives

Reviewer 2: ```suggestion
created. ``schedule_interval`` is defined as a DAG argument, and receives
```

preferably a
`cron expression <https://en.wikipedia.org/wiki/Cron#CRON_expression>`_ as
a ``str``, or a ``datetime.timedelta`` object. Alternatively, you can also

> use one of these cron "preset":

Reviewer 2: ```suggestion
use one of these cron "presets":
```

will be triggered soon after 2020-01-01T23:59. In other words, the job instance is 
started once the period it covers has ended.  The execution_date passed in the dag 
will also be 2020-01-01.

The first ``DAG Run`` is created based on the minimum ``start_date`` for the tasks in your DAG. 
Subsequent ``DAG Runs`` are created by the scheduler process, based on your DAG’s ``schedule_interval``, 
sequentially. If your start_date is 2020-01-01 and schedule_interval is @daily the first run 
will be created on 2020-01-02 i.e. after your start date has passed.

Re-run DAG
''''''''''
There can be cases where you will want to execute your DAG again. One such case is when the scheduled
DAG run fails. Another can be the scheduled DAG run wasn't executed due to low resources or the DAG being turned off.

Catchup
-------

An Airflow DAG with a ``start_date``, possibly an ``end_date``, and a ``schedule_interval`` defines a 
series of intervals which the scheduler turn into individual DAG Runs and execute. A key capability 
of Airflow is that these DAG Runs are atomic and idempotent items. The scheduler, by default, will

> kick off a DAG Run for any interval that has not been run (or has been cleared). This concept is called Catchup.

Reviewer 2: This is mostly true, but slightly misleading in a few edge cases. I'm not sure if this is worth mentioning here or not.

Catchup, and the scheduler in general will not "fill in gaps" - it will only look forward from the most recent dag run.

For example:

- I have a daily dag with catchup=False running. This is "d0"
- I pause that dag for 3 days
- I then start it again.

At this point we have dagruns for d0, d4, d5, ...

- I edit the dag to set catchup=True

The scheduler will not go and "fill in" d1, d2, d3.

Author: Updated with changes.

If your DAG is written to handle its own catchup (i.e. not limited to the interval, but instead to ``Now`` for instance.), 
then you will want to turn catchup off. This can be done by setting ``catchup = False`` in DAG  or ``catchup_by_default = False``
in configuration file. When turned off, the scheduler creates a DAG run only for the latest interval.

.. code:: python

    """
    Code that goes along with the Airflow tutorial located at:
    https://github.com/apache/airflow/blob/master/airflow/example_dags/tutorial.py
    """
    from airflow import DAG
    from airflow.operators.bash_operator import BashOperator
    from datetime import datetime, timedelta

    default_args = {
        'owner': 'Airflow',
        'depends_on_past': False,

>         'start_date': datetime(2015, 12, 1),

Reviewer 2: I know this is what we do in most cases in our docs, but can you move start_date out of `default_args` and in to just a DAG arg please?

        'email': ['airflow@example.com'],
        'email_on_failure': False,
        'email_on_retry': False,
        'retries': 1,
        'retry_delay': timedelta(minutes=5)
    }

    dag = DAG(
        'tutorial',
        default_args=default_args,
        description='A simple tutorial DAG',
        schedule_interval='@daily',
        catchup=False)

In the example above, if the DAG is picked up by the scheduler daemon on 2016-01-02 at 6 AM, 
(or from the command line), a single DAG Run will be created, with an `execution_date` of 2016-01-01, 
and the next one will be created just after midnight on the morning of 2016-01-03 with an execution date of 2016-01-02.

If the ``dag.catchup`` value had been True instead, the scheduler would have created a DAG Run 
for each completed interval between 2015-12-01 and 2016-01-02 (but not yet one for 2016-01-02, 
as that interval hasn’t completed) and the scheduler will execute them sequentially. 

Catchup is also triggered when you turn off a DAG for a specified period of time and then re-enable.

This behavior 
is great for atomic datasets that can easily be split into periods. Turning catchup off is great 

> if your DAG Runs perform backfill internally.

Reviewer 2: ```suggestion
if your DAG performs catchup internally.
```

Backfill
---------
There can be the case when you may want to run the dag for a specified historical period e.g. a data pipeline
which dumps data in a DFS every day and another pipeline which requires last 1 month of data in DFS. 
This is known as Backfill.

You may want to backfill the data even in the cases when catchup is disabled. This can be done through CLI. 
Run the below command

.. code:: bash

    airflow backfill -s START_DATE -e END_DATE dag_id

The above command will re-run all the instances of the dag_id for all the intervals within the start date and end date.

Re-run Tasks
------------
It can happen, that some of the tasks can fail during the scheduled run. Once you have fixed 
the errors after going through the logs, you can re-run the tasks by clearing it for the 

> scheduled date. Clearing a task instance doesn't delete the task instance record. 

Reviewer 2: The other thing that I like to do sometimes is to use the "Run"+"Ignore Task state" buttons -- that way you end up with "Attempt 3 of 2" which makes it clearer to me that it is a retry. I think this doesn't help down-stream tasks get rescheduled though as they would already be in upstream failed.

(I basically don't like how "destructive" Clear is for history.)

Instead it updates ``max_tries`` to ``0`` and set the current task instance state to be ``None``, this forces the task to re-run.

> Select the failed task and click on **Clear**. This will clear the status of the task from 

Reviewer 2: It's not clear from this where the Clear button is (I know what you mean, it's just not clear in this doc).

```suggestion
Select the failed task and click on **Clear** in the Task Instance modal. This will clear the status of the task from 
```

maybe? Or 
```suggestion
Click on the failed task in the Tree or Graph views and then click on **Clear**.
``

failed to ``None`` and the executor will re-run it.

There are multiple options you can select to re-run - 

* Past - All the instances of the task in the  runs before the current DAG's execution date
* Future -  All the instances of the task in the  runs after the current DAG's execution date
* Upstream - The upstream tasks in the current DAG
* Downstream - The downstream tasks in the current DAG
* Recursive - All the tasks in the child DAGs and parent DAGs
* Failed - Only the failed tasks in the current DAG

You can also clear the task through CLI using the command:

.. code:: bash

    airflow tasks clear dag_id -t task_regex -s START_DATE -d END_DATE

This will clear all instances of the tasks matching the regex for the dag_id which have run during 
the mentioned interval. For more options, you can run the command:

.. code:: bash

    airflow tasks clear -h

**Note**: When clearing a set of tasks’ state in hope of getting them to re-run, it is important 
to keep in mind the DAG Run’s state too as it defines whether the scheduler should look

> into triggering tasks for that run.

Reviewer 2: I think that if you clear tasks it will also set the dagrun state to running? But I'm not certain of that.

External Triggers
'''''''''''''''''

Note that ``DAG Runs`` can also be created manually through the CLI. Just run the command -

.. code:: bash

    airflow dags trigger -e execution_date run_id

> The ``DAG Runs`` created externally to the scheduler get associated to the trigger’s timestamp, and will be displayed 

Reviewer 2: ```suggestion
The DAG Runs created externally to the scheduler get associated to the trigger’s timestamp, and will be displayed 
```

We probably don't need to make DAG Runs be monospaced, capitalized is enough.

Reviewer 2: (This applies to the whole document, not just here)

+--------------+----------------------------------------------------------------+---------------+
| ``@daily``   | Run once a day at midnight                                     | ``0 0 * * *`` |
+--------------+----------------------------------------------------------------+---------------+
| ``@weekly``  | Run once a week at midnight on Sunday morning                  | ``0 0 * * 0`` |
+--------------+----------------------------------------------------------------+---------------+
| ``@monthly`` | Run once a month at midnight of the first day of the month     | ``0 0 1 * *`` |
+--------------+----------------------------------------------------------------+---------------+
| ``@yearly``  | Run once a year at midnight of January 1                       | ``0 0 1 1 *`` |
+--------------+----------------------------------------------------------------+---------------+

Your DAG will be instantiated for each schedule along with a corresponding 
DAG Run entry in backend.

.. note::

    If you run a DAG on a schedule_interval of one day, the run stamped 2020-01-01 
    will be triggered soon after 2020-01-01T23:59. In other words, the job instance is 

>     started once the period it covers has ended.  The execution_date passed in the dag 

Reviewer 2: ```suggestion
    started once the period it covers has ended.  The ``execution_date`` available in the context
```

    will also be 2020-01-01.

    The first DAG Run is created based on the minimum ``start_date`` for the tasks in your DAG. 
    Subsequent DAG Runs are created by the scheduler process, based on your DAG’s ``schedule_interval``, 
    sequentially. If your start_date is 2020-01-01 and schedule_interval is @daily the first run 
    will be created on 2020-01-02 i.e. after your start date has passed.

Re-run DAG
''''''''''
There can be cases where you will want to execute your DAG again. One such case is when the scheduled

> DAG run fails. Another can be the scheduled DAG run wasn't executed due to low resources or the DAG being turned off.

Reviewer 2: Low resources shouldn't stop a dag run from being created, so lets remove that bit.

.. _dag-catchup:

Catchup
-------

An Airflow DAG with a ``start_date``, possibly an ``end_date``, and a ``schedule_interval`` defines a 

> series of intervals which the scheduler turn into individual DAG Runs and execute. A key capability 

Reviewer 2: ```suggestion
series of intervals which the scheduler turns into individual DAG Runs and then executes. A key capability 
```

> of Airflow is that these DAG Runs are atomic and idempotent items. The scheduler, by default, will

Reviewer 2: ```suggestion
of Airflow is that these DAG Runs should be atomic and idempotent items. The scheduler, by default, will
```

(should be, as Airflow doesn't do this magically, but it depends on how the tasks are written.)

Author: Ok. then I guess I should remove the line only as it doesn't make much sense.

Author: or I can add The runs should be atomic and idempotent for the catchup to function as expected.

.. code:: python

    """
    Code that goes along with the Airflow tutorial located at:
    https://github.com/apache/airflow/blob/master/airflow/example_dags/tutorial.py
    """
    from airflow import DAG
    from airflow.operators.bash_operator import BashOperator
    from datetime import datetime, timedelta

    default_args = {
        'owner': 'Airflow',
        'depends_on_past': False,
        'email': ['airflow@example.com'],
        'email_on_failure': False,
        'email_on_retry': False,
        'retries': 1,
        'retry_delay': timedelta(minutes=5)
    }

    dag = DAG(
        'tutorial',
        default_args=default_args,
        start_date=datetime(2015, 12, 1),
        description='A simple tutorial DAG',
        schedule_interval='@daily',
        catchup=False)

In the example above, if the DAG is picked up by the scheduler daemon on 2016-01-02 at 6 AM, 
(or from the command line), a single DAG Run will be created, with an `execution_date` of 2016-01-01, 
and the next one will be created just after midnight on the morning of 2016-01-03 with an execution date of 2016-01-02.

> If the ``dag.catchup`` value had been True instead, the scheduler would have created a DAG Run 

Reviewer 5: ```suggestion
If the ``dag.catchup`` value had been ``True`` instead, the scheduler would have created a DAG Run 
```

for each completed interval between 2015-12-01 and 2016-01-02 (but not yet one for 2016-01-02, 
as that interval hasn’t completed) and the scheduler will execute them sequentially. 

Catchup is also triggered when you turn off a DAG for a specified period of time and then re-enable.

This behavior 
is great for atomic datasets that can easily be split into periods. Turning catchup off is great 
if your DAG performs catchup internally.

Backfill
---------
There can be the case when you may want to run the dag for a specified historical period e.g. a data pipeline

> which dumps data in a DFS every day and another pipeline which requires last 1 month of data in DFS. 

Reviewer 2: I'm not sure this is the best example of backfill.

Backfill's primary aim is for running for date periods before the start date of the task. (Otherwise the scheduler would see the date is before start_date and not do anything)

This is known as Backfill.

You may want to backfill the data even in the cases when catchup is disabled. This can be done through CLI. 
Run the below command

.. code:: bash

    airflow backfill -s START_DATE -e END_DATE dag_id

> The above command will re-run all the instances of the dag_id for all the intervals within the start date and end date.

Reviewer 2: We should link to https://airflow.apache.org/cli.html#backfill

Re-run Tasks
------------
It can happen, that some of the tasks can fail during the scheduled run. Once you have fixed 
the errors after going through the logs, you can re-run the tasks by clearing it for the 
scheduled date. Clearing a task instance doesn't delete the task instance record. 
Instead it updates ``max_tries`` to ``0`` and set the current task instance state to be ``None``, this forces the task to re-run.

Click on the failed task in the Tree or Graph views and then click on **Clear**.
``

> failed to ``None`` and the executor will re-run it.

Reviewer 2: Odd line break here.

There are multiple options you can select to re-run - 

> * Past - All the instances of the task in the  runs before the current DAG's execution date

Reviewer 5: ```suggestion
* **Past** - All the instances of the task in the  runs before the current DAG's execution date
```

We should probably bold all the options

* Future -  All the instances of the task in the  runs after the current DAG's execution date
* Upstream - The upstream tasks in the current DAG
* Downstream - The downstream tasks in the current DAG
* Recursive - All the tasks in the child DAGs and parent DAGs
* Failed - Only the failed tasks in the current DAG

You can also clear the task through CLI using the command:

.. code:: bash

    airflow tasks clear dag_id -t task_regex -s START_DATE -d END_DATE

This will clear all instances of the tasks matching the regex for the dag_id which have run during 
the mentioned interval. For more options, you can run the command:

.. code:: bash

>     airflow tasks clear -h

Reviewer 2: Link to cli docs here too.

Author: Should I remove the example and just have the CLI doc link or put CLI docs link below the example?

External Triggers
'''''''''''''''''

Note that DAG Runs can also be created manually through the CLI. Just run the command -

.. code:: bash

    airflow dags trigger -e execution_date run_id

The DAG Runs created externally to the scheduler get associated to the trigger’s timestamp, and will be displayed 

> in the UI alongside scheduled DAG runs. The execution date passed inside the DAG can be specified using -e argument 

Reviewer 5: ```suggestion
in the UI alongside scheduled DAG runs. The execution date passed inside the DAG can be specified using ``-e`` argument 
```
